src/piadm.sh

```diff
@@ -197,129 +195,126 @@ mount_installmedia() {
 # XXX WARNING - there is a security discussion to be had about the integrity
 # of the source.
 install() {
-	piname_present_get_bootfs $1 $2
+	piname_present_get_bootfs "$1" "$2"
 	tdir=$(mktemp -d)
-	mkdir ${tdir}/mnt
+	mkdir "${tdir}/mnt"
 
 	# $1 contains a "source".  Deal with it correctly in the big
 	# if/elif/else block.  Once done, we can copy over bits into $tdir or
 	# ${tdir}/mnt.
-	# 
+	#
 
 	# Special-case of "latest"
 	if [[ "$1" == "latest" ]]; then
 		# Well-known URL for the latest PI using conventions from
 		# URL_PREFIX.  Grab the latest-version ISO. Before proceeding,
 		# make sure it's the current one.
 		iso=yes
-		${CURL} -o ${tdir}/smartos.iso ${URL_PREFIX}/smartos-latest.iso
-		if [[ $? -ne 0 ]]; then
-			code=$?
-			/bin/rm -rf ${tdir}
+		"${CURL[@]}" -o "${tdir}/smartos.iso" "${URL_PREFIX}/smartos-latest.iso"
+		code=$?
+		if [[ $code -ne 0 ]]; then
+			/bin/rm -rf "${tdir}"
 			fatal "Curl exit code $code"
 		fi
-		mount -F hsfs ${tdir}/smartos.iso ${tdir}/mnt
+		mount -F hsfs "${tdir}/smartos.iso" "${tdir}/mnt"
 
 		# For now, assume boot stamp and PI stamp are the same on an ISO...
-		stamp=$(cat ${tdir}/mnt/etc/version/boot)
+		stamp=$(cat "${tdir}/mnt/etc/version/boot")
 	elif [[ "$1" == "media" ]]; then
 		# Scan the available media to find what we seek.  Same advice
 		# about making sure it's the current one.
 		iso=yes
-		mount_installmedia ${tdir}/mnt
-		if [[ $? -ne 0 ]]; then
-			/bin/rm -rf ${tdir}
+		if ! mount_installmedia "${tdir}/mnt" ; then
+			/bin/rm -rf "${tdir}"
 			err "Cannot find install media"
 		fi
 
 		# For now, assume boot stamp and PI stamp are the same on
 		# install media.
-		stamp=$(cat ${tdir}/mnt/etc/version/boot)
+		stamp=$(cat "${tdir}/mnt/etc/version/boot")
 	elif [[ -f $1 ]]; then
 		# File input!  Check for what kind, etc. etc.
 
 		# WARNING:  Depends GREATLY on the output of file(1)
-		filetype=$(file $1 | awk '{print $2}')
+		filetype=$(file "$1" | awk '{print $2}')
 		if [[ "$filetype" == "ISO" ]]; then
 			# Assume .iso file.
 			iso=yes
-			mount -F hsfs $1 ${tdir}/mnt
-			stamp=$(cat ${tdir}/mnt/etc/version/boot)
+			mount -F hsfs "$1" "${tdir}/mnt"
+			stamp=$(cat "${tdir}/mnt/etc/version/boot")
 		elif [[ "$filetype" == "gzip" ]]; then
 			# SmartOS PI.  Let's confirm it's actually a .tgz...
-			gtar -xzOf $1 > /dev/null 2>&1
-			if [[ $? -ne 0 ]]; then
-				/bin/rm -rf ${tdir}
+
+			if ! gtar -xzOf "$1" > /dev/null 2>&1; then
+				/bin/rm -rf "${tdir:?}"
 				err "File $1 is not an ISO or a .tgz file."
 			fi
 			# We're most-likely good here.  NOTE: SmartOS/Triton
 			# PI files expand to platform-$STAMP.  Fix it here
 			# before proceeding.
-			gtar -xzf $1 -C ${tdir}/mnt
-			mv ${tdir}/mnt/platform-* ${tdir}/mnt/platform
+			gtar -xzf "$1" -C "${tdir}/mnt"
+			mv "${tdir}/mnt/platform-*" "${tdir}/mnt/platform"
```

Comment:
This breaks because I know that a freshly-extracted tarball will have platform-* in it, but I have no idea what * expands to (what stamp is in the .tgz's platform) until below.  If that breaks shellcheck, you'll need to extract the stamp from platform-* (i.e. expand * yourself).  Seems silly when shell does a great job of expansion already.  And we're guaranteed (unless it's a malicious tarball) that there's only one platform-* expansion.

Reply:
fixed the quoting